File: 2021/2021W41/preppindataW41.R
library(readr)
library(dplyr)
library(tidyr)
library(stringr)

final <- read_table("C:/Data/PreppinData/Southend Stats.csv") %>%
  rename('Pts' = names(.)[length(names(.))-1]) %>%
  arrange(`SEASON`) %>%
  mutate('Special Circumstances'=if_else(`SEASON`==max(`SEASON`),'Incomplete',
                                         if_else(substr(`SEASON`,1,4)=='1939','Abandoned due to WW2','N/A')),
         'POS' = if_else(`Special Circumstances`=='N/A',`POS`,NA_character_),
         'League Num' = if_else(`LEAGUE`=='FL-CH',0,
                                if_else(`LEAGUE`=='NAT-P',5, 
                                        as.double(str_extract_all(`LEAGUE`,'(\\d)')))),
         'Outcome' = if_else(`League Num` - lead(`League Num`) >0,'Promoted',
                             if_else(`League Num` - lead(`League Num`) <0,'Relegated','Same League')),
         'Year' = as.integer(substr(`SEASON`,1,4))) %>%
  complete(., `Year` = full_seq(`Year`, period = 1)) %>%
  mutate('Special Circumstances' = if_else(is.na(`SEASON`),if_else(`Year`<1939,'WW1','WW2'),`Special Circumstances`),
         'SEASON' = if_else(is.na(`SEASON`),paste(`Year`,(`Year`+1)%%100,sep='-'),`SEASON`)) %>%
  select(c('SEASON', 'Outcome', 'Special Circumstances', 'LEAGUE', 'P', 'W', 'D', 'L', 'F', 'A', 'Pts', 'POS'))

View(final)

File: 2021/2021W41/PreppindataW41.py
import pandas as pd
import numpy as np
import re

stats = pd.read_csv("C:\\Data\\PreppinData\\Southend Stats.csv", sep='\s+')
stats = stats.rename(columns={stats.columns[-2]:'Pts'})
stats= stats.sort_values(by='SEASON')
stats['Special Circumstances'] = stats.apply(lambda x: 'Incomplete' if x['SEASON']==stats['SEASON'].max() else 
                                             'Abandoned due to WW2' if x['SEASON'][0:4]=='1939' else 'N/A', axis=1)
stats['POS'] = stats.apply(lambda x: x['POS'] if x['Special Circumstances']=='N/A' else None, axis=1)
stats['League Num'] = stats['LEAGUE'].apply(lambda x: 0 if x=='FL-CH' else 5 if x=='NAT-P' else int(re.search('(\d)',x).group(1)))
stats['Outcome Key'] = stats['League Num'] - stats['League Num'].shift(-1) 
stats['Outcome'] = stats['Outcome Key'].apply(lambda x: 'Promoted' if x>0 else 'Relegated' if x<0 else 'Same League' if x==0 else None)

stats.index = stats['SEASON'].apply(lambda x: int(x[0:4]))
stats = stats.reindex(np.arange(stats.index.min(), stats.index.max() + 1)).rename_axis('Year').reset_index()
stats['SEASON'] = stats['Year'].apply(lambda x: f'{x}-{str(x+1)[-2:]}')
stats['Special Circumstances'] = stats.apply(lambda x: x['Special Circumstances'] if pd.notna(x['Special Circumstances']) else
                                             'WW1' if x['Year']<1939 else 'WW2', axis=1)
stats = stats[['SEASON', 'Outcome', 'Special Circumstances', 'LEAGUE', 'P', 'W', 'D', 'L', 'F', 'A', 'Pts', 'POS']]
